clang/lib/Sema/Sema.cpp

```diff
@@ -1533,6 +1540,24 @@ void Sema::markKnownEmitted(
       }
     }
 
+    // Function object calls are not walked above.
```

Comment:
Could you please explain why "Function object calls are not walked above"?
I don't see anything specific for function objects in the following code?

Reply:
The function object call expression appears in the IR as:
CXXMethodDecl 0x55fbe8077ef0 <exc_func.cpp:18:44, col:55> col:41 operator() 'void () const' inline
`-CompoundStmt 0x55fbe8078040 <col:46, col:55>
  `-CallExpr 0x55fbe8078020 <col:48, col:52> 'void'
    `-ImplicitCastExpr 0x55fbe8078008 <col:48> 'void (*)()' <FunctionToPointerDecay>
      `-DeclRefExpr 0x55fbe8077fe8 <col:48> 'void ()' lvalue Function 0x55fbe804d308 'bar' 'void ()'

This kind of call is not in the Call Graph created for the device, so I need to walk it here.   It is the absence of this that makes the exception code such as in the example test added, not be diagnosed.   This is what the code below this comment does.   I am not sure what you mean by "I don't see anything specific".   The code below, basically the block below "if (CS)", extracts the underlying function from the call expression.

Reply:
Okay, I understand, thank you very much for the explanation.
Do you know why the such calls are missing in the device Call Graph?

Reply:
> Okay, I understand, thank you very much for the explanation.
> Do you know why such calls are missing in the device Call Graph?

I really don't know.  That's the honest answer.   :-)
This code was leveraged from existing implementation for CUDA; perhaps this does not make sense there?  On the other hand, I do not see any explicit comments stating that this is deliberate.

Reply:
> I really don't know. That's the honest answer. :-)

I think it would be great if we understand the problem before making any changes, isn't it? :)
AFAIK, more or less modern versions of CUDA support lambda functions.

Reply:
CallGraph was originally implemented for some diagnosis components and some clang-tidy checks, so it misses a whole bunch of situations and is surprisingly undertested and under-maintained.  I went through in community last year and had to enable a number of features to it for exactly this reason.

I'm not surprised that it gets confused quickly by lambdas/function objects.

Reply:
I agree with Erich's comment. I saw a bunch of instances where the CallGrph isn't complete, for example calls to user-defined operators.

Reply:
Shouldn't we improve the CallGraph in this case?

Reply:
I feel we should do that a separate generic task; I can add a "FIX ME" in front of this change if/when someone gets to doing the general fix.  I am worried about the things that might break with a general change in the call-graph, particularly if it is under-tested as Erich claims above.

Reply:
Okay, I will investigate fixing the call-graph and report back in the next few days.

Reply:
This doesn't actually use the clang::CallGrpah like we were discussing above.  "DeviceCallGraph" is simply a dense map.  

I see that OpenMP and CUDA do similar things, do they have the same problem?  If so, is there a more generic solution for this?

Reply:
Alright, I dug in.  This isn't the solution.  The problem is that CheckSYCLCall is getting the caller incorrectly.

getCurFunctionDecl gets the wrong thing.  It will get the function containing the lambda at definition time, rather than the lambda operator().  In CodeGen there is a fantastic alternative here that is about a 6 character replacement (I think it is getCurBlockDecl()), but it isn't exactly as easy here. 

I believe the best way (after looking through SEMA) is to change it to getCurLexicalContext, then check if that DeclContext is a FunctionOrMethod.  If so, use it as the Caller FunctionDecl.

That said, that brings up the following issue:

    int foo() { throw 1; }
    struct S {
        int bar = foo();
    };
    int main() { kernel_single_task<class fake_kernel>([](){ S s; });

That ALSO has the problem, except we do our reporting at a functiondecl->caller level, when the declcontext can actually be a CXXRecordDecl as well.  Presumably, the devicecallgraph needs to have a NamedDecl (or even a Decl?, heck, perhaps a DeclContext?!) as its caller, since functions can be called at namespace, global, or class scope.

Consider:

int foo() { throw 1; }
int global_var = foo(); // always executed, even on device.  Probably needs to diagnose as well!

Reply:
Alright, I dug in.  This isn't the solution.  The problem is that CheckSYCLCall is getting the caller incorrectly.

getCurFunctionDecl gets the wrong thing.  It will get the function containing the lambda at definition time, rather than the lambda operator().  In CodeGen there is a fantastic alternative here that is about a 6 character replacement (I think it is getCurBlockDecl()), but it isn't exactly as easy here. 

I believe the best way (after looking through SEMA) is to change it to getCurLexicalContext, then check if that DeclContext is a FunctionOrMethod.  If so, use it as the Caller FunctionDecl.

That said, that brings up the following issue:

    int foo() { throw 1; }
    struct S {
        int bar = foo();
    };
    int main() { kernel_single_task<class fake_kernel>([](){ S s; });

That ALSO has the problem, except we do our reporting at a functiondecl->caller level, when the declcontext can actually be a CXXRecordDecl as well.  Presumably, the devicecallgraph needs to have a NamedDecl (or even a Decl?, heck, perhaps a DeclContext?!) as its caller, since functions can be called at namespace, global, or class scope.

Consider:

    int foo() { throw 1; }
    int global_var = foo(); // always executed, even on device.  Probably needs to diagnose as well!





Ugg... Thought of another example, so I'm not sure LEXICAL context is the right one.  Experimentation is needed.

    struct S {
        static int f;
    };
    S::f = foo(); // always executed, but lexical context is namespace/TU decl.  Really, it should be part of 'S'.  

I wonder if we need to do some work later to get the actual declaration context in those cases.

Reply:
Thanks Erich!   The test cases are illuminating!
I will take a cut a redesigning how we do the device diagnostics.